crates/sui-adapter/src/programmable_transactions/context.rs

```diff
@@ -259,17 +302,247 @@ where
+        self.results
+            .push(results.into_iter().map(ResultValue::new).collect());
         Ok(())
     }
 
+    /// Determine the object changes and collect all user events
+    pub fn finish(self) -> Result<ExecutionResults, ExecutionError> {
+        use sui_types::error::convert_vm_error;
+        let Self {
+            protocol_config,
+            vm,
+            state_view,
+            tx_context,
+            gas_status,
+            session,
+            additional_transfers,
+            new_packages,
+            gas,
+            mut inputs,
+            results,
+            user_events,
+            ..
+        } = self;
+        let tx_digest = tx_context.digest();
+        let sender = tx_context.sender();
+        let mut additional_writes = BTreeMap::new();
+        let mut input_object_metadata = BTreeMap::new();
+        // check for unused inputs
+        // track by-ref input objects as being written
+        inputs.push(gas);
+        for input in inputs {
+            let InputValue {
+                object_metadata: object_metadata_opt,
+                inner:
+                    ResultValue {
+                        last_usage_kind,
+                        value,
+                    },
+            } = input;
+            if last_usage_kind.is_none() {
+                panic!("unused input")
+            }
+            let Some(object_metadata) = object_metadata_opt else { continue };
+            let Some(Value::Object(object_value)) = value else { continue };
+            if object_metadata.is_mutable_input {
+                add_additional_write(&mut additional_writes, object_metadata.owner, object_value);
+            }
+            input_object_metadata.insert(object_metadata.id, object_metadata);
+        }
+        // check for unused values
+        for (i, command_result) in results.iter().enumerate() {
+            for (j, result_value) in command_result.iter().enumerate() {
+                let ResultValue {
+                    last_usage_kind,
+                    value,
+                } = result_value;
+                match value {
+                    None => (),
+                    Some(Value::Object(_)) => panic!("unused value without drop {i} {j}"),
+                    Some(Value::Raw(ValueType::Any, _)) => (),
+                    Some(Value::Raw(ValueType::Loaded { abilities, .. }, _)) => {
+                        // - nothing to check for drop
+                        // - if it does not have drop, but has copy,
+                        //   the last usage must be a take/clone in order to "lie" and say that the
+                        //   last usage is actually a take instead of a clone
+                        // - Otherwise, an error
+                        if abilities.has_drop() {
+                        } else if abilities.has_copy()
+                            && !matches!(last_usage_kind, Some(UsageKind::Take | UsageKind::Clone))
```

Comment:
Out of interest, why distinguish between `Take` and `Clone`?  IIRC, the semantics were that there was only `Take`, but when you `Take` an object with `copy`, it was cloned, and here you could just if the last use was a take.

Reply:
I'm not sure I understand the question. `Take` sets the value to `None`. `Clone` does not. 

Really the only last usage kind here will be `Clone` since it has `copy`. But from a rules standpoint, we just need to check that the last usage was by value. And that means either Take or Clone

Reply:
The suggestion was to have only `Take`, but that if you call `Take` on a value that has `copy` then it doesn't set the value to `None`, and here, if the value has `copy` then it's fine its last usage was a take.